### PR TITLE
feat: add api endpoint to fetch messages

### PR DESCRIPTION
**Pull request by erolkskn**
## Related issue(s)
#2639 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I am following the [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security. vulnerability, 
      I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
@erolkskn let me know when this is good for a review :)

**Comment by erolkskn** [comment]
@aeneasr Hi aeneas, i think it's ready for review now :) Btw, about the last commit, autogen bot somehow formatted swagger doc blocks wrong (see: 4159b93ae3f8175cf7ccf77d34e4a7a2d0181d4f). That commit caused swagger to be not able to generate spec. So i needed to make that change in order to make sdk generation work again (see: https://github.com/ory/kratos/pull/2651/commits/b51a916fd67385ae8f4c44f0ce654c33e720711c).

**Comment by aeneasr** [comment]
Great, @erolkskn ! Yeah the problem is that Go 1.19 changed the formatter, which now breaks the go-swagger generation. I fixed the issues on master, so if you merge / rebase they should be resolved.

@jonas-jonas can you please take a look at the PR? :)

Also, we probably need a bit of documentation for this feature, maybe in a new doc in https://github.com/ory/docs/tree/master/docs/kratos/emails-sms ?

File: persistence/sql/persister_courier.go
Comment (jonas-jonas) [inline persistence/sql/persister_courier.go:32]: I think, this should be `corp.ContextualizeNID(ctx, p.nid)` to keep it consistent with the other methods in the persister.

File: persistence/sql/persister_courier.go
Comment (jonas-jonas) [inline persistence/sql/persister_courier.go:43]: I know this wasn't part of the issue, but seeing this line made me think of this: It might be worth adding a sort-direction to the `courier.MessagesFilter` struct to set the direction (`DESC` or `ASC`) here. This could be used to get the first sent emails matching the other filters. cc @aeneasr

**Review by jonas-jonas — CHANGES_REQUESTED** [review]
Very nice work! Thank you for your contribution. This looks pretty good so far, I just have two ideas on how to improve it.

File: persistence/sql/persister_courier.go
Comment (erolkskn) [reply persistence/sql/persister_courier.go:32]: Yeah it was actually like that before, but after i rebased this branch onto master yesterday, i seen that `corp.ContextualizeNID(ctx, pid)` is refactored to `p.NetworkID(ctx)` in commit b0d8f3853886228a64e82437643a82b3970d6ff7. That's why i did go with `p.NetworkID(ctx)`

Comment (jonas-jonas) [reply persistence/sql/persister_courier.go:32]: Huh, interesting, I missed that—sorry about that. This looks good to me then.

File: persistence/sql/persister_courier.go
Comment (erolkskn) [reply persistence/sql/persister_courier.go:43]: @jonas-jonas I think that's a good idea. Should i proceed to implement this now ? Or are we waiting for @aeneasr 's approval ?

Comment (aeneasr) [reply persistence/sql/persister_courier.go:43]: I agree that it would be useful, but since we do not have ordering yet in our REST APIs and we have no design document describing how such ordering would look like, i'd like to defer this for another time :)

Comment (jonas-jonas) [reply persistence/sql/persister_courier.go:43]: Then LGTM.

**Review by jonas-jonas — DISMISSED** [review]
LGTM!

File: persistence/sql/persister_test.go
Comment (aeneasr) [inline persistence/sql/persister_test.go:243]: We use this helper elswhere so it's needed to have this little function. Can you please revert those changes? Thanks! :)

**Review by aeneasr — CHANGES_REQUESTED** [review]
Thank you! This looks great, just one topic we need to change :)

File: persistence/sql/persister_test.go
Comment (erolkskn) [reply persistence/sql/persister_test.go]: @aeneasr Alright, i reverted the change. How does it look now ? :)

**Review by aeneasr — COMMENTED** [review]
It looks fantastic! I will do another pass next week but I think this is good to be merged :)

File: courier/handler.go
Comment (zepatrik) [inline courier/handler.go:96]: Shouldn't this be 
```suggestion
	if !h.r.Config().IsInsecureDevMode(r.Context()) {
```

File: courier/handler_test.go
Comment (zepatrik) [inline courier/handler_test.go:80]:
```suggestion
			messages[i].Type = courier.MessageTypeEmail
			messages[i].Body = "body content"
```

File: courier/handler_test.go
Comment (zepatrik) [inline courier/handler_test.go:166]:
```suggestion
						assert.Equal(t, "body content", item.Get("body").String())
```

File: courier/handler_test.go
Comment (zepatrik) [inline courier/handler_test.go:179]:
```suggestion
						assert.Equal(t, "<redacted-unless-dev-mode>", item.Get("body").String())
```

File: courier/message.go
Comment (zepatrik) [inline courier/message.go:43]:
```suggestion
	switch s := stringsx.SwitchExact(str); {
	case s.AddCase(MessageStatusQueued.String()):
		return MessageStatusQueued, nil
	case s.AddCase(MessageStatusSent.String()):
		return MessageStatusSent, nil
	case s.AddCase(MessageStatusProcessing.String()):
		return MessageStatusProcessing, nil
	case s.AddCase(MessageStatusAbandoned.String()):
		return MessageStatusAbandoned, nil
	default:
		return 0, errors.WithStack(herodot.ErrBadRequest.WithError(s.ToUnknownCaseErr()).WithReason("Message status is not valid"))
	}
```

Comment (zepatrik) [reply courier/message.go:43]: Using `herodot` errors automatically sets the status code correctly.

File: courier/message.go
Comment (zepatrik) [inline courier/message.go:118]: Similar to above suggestion :wink:

File: courier/message_test.go
Comment (zepatrik) [inline courier/message_test.go:12]:
```suggestion
	require.ErrorIs(t, invalid.IsValid(), herodot.ErrBadRequest, "IsValid() should return an error when message status is invalid")
```

File: courier/message_test.go
Comment (zepatrik) [inline courier/message_test.go:31]:
```suggestion
		require.ErrorIs(t, result.IsValid(), herodot.ErrBadRequest)
```

File: courier/message_test.go
Comment (zepatrik) [inline courier/message_test.go:37]:
```suggestion
	require.ErrorIs(t, invalid.IsValid(), herodot.ErrBadRequest, "IsValid() should return an error when message type is invalid")
```

File: courier/message_test.go
Comment (zepatrik) [inline courier/message_test.go:54]:
```suggestion
		require.ErrorIs(t, result.IsValid(), herodot.ErrBadRequest)
```

File: courier/test/persistence.go
Comment (zepatrik) [inline courier/test/persistence.go:124]:
```suggestion
			assert.Equal(t, messages[len(messages)-1].ID, ms[0].ID)
			assert.ElementsMatch(t, messages, ms)
```
`assert.ElementsMatch` ignores the order.`

File: courier/message.go
Comment (zepatrik) [inline courier/message.go:29]:
```suggestion
// A Message's Status
//
// swagger:enum messageStatus
type messageStatus string

const (
	messageStatusQueuedText messageStatus     = "queued"
	messageStatusSentText messageStatus       = "sent"
	messageStatusProcessingText messageStatus = "processing"
	messageStatusAbandonedText messageStatus  = "abandoned"
)
```

You might be able to avoid the patches if you do it this way. What is important is that the type name is the same as the `// swagger:enum name`, and that all values are in one block immediately after the type for this to work. But you can also just keep the patch if this does not work properly.

File: .schema/openapi/patches/courier.yaml
Comment (zepatrik) [inline .schema/openapi/patches/courier.yaml:24]: Maybe make this just one op?
```suggestion
- op: replace
  path: /components/schemas/messageStatus
  value:
    type: string
    enum:
      - queued
      - sent
      - processing
      - abandoned
# Makes messageType a string enum
- op: replace
  path: /components/schemas/messageType
  value:
    type: string
    enum:
      - email
      - phone
```
That way it is exactly what you want.

**Review by zepatrik — CHANGES_REQUESTED** [review]
Nice, just found some minor bugs :wink: 
But overall nearly there :+1:

File: courier/handler.go
Comment (erolkskn) [reply courier/handler.go:96]: Most definitely!

File: courier/test/persistence.go
Comment (erolkskn) [reply courier/test/persistence.go:124]: Didn't know that, thank you :-)

File: .schema/openapi/patches/courier.yaml
Comment (erolkskn) [reply .schema/openapi/patches/courier.yaml:24]: Yeah that's way better approach, idk what i was thinking writing that piece of code :-)

File: courier/message.go
Comment (erolkskn) [reply courier/message.go:43]: @zepatrik I guess `ErrBadRequest.WithError` should be `ErrBadRequest.WithWrap` right ?

File: courier/test/persistence.go
Comment (erolkskn) [reply courier/test/persistence.go:124]: @zepatrik On a second thought, this can't work on this scenario. Test cases above the `case=list messages` either updates the messages with `SetMessageStatus` or `IncrementMessageSendCount` which none of them takes a reference to the `courier.Message` itself. So `ListMessages` method always returns messages with different `MessageStatus` or `SendCount` than what we have in `messages` variable. So `assert.ElementsMatch` statement bound to be failed due to the reasons i listed above.

File: courier/message.go
Comment (erolkskn) [reply courier/message.go:29]: @zepatrik That would definitely make us avoid the patches, but is it really necessary to introduce another type just for this ? I think that might complicate code-base a little bit more (all the need for casting `messageStatus` to `string` and vice versa, new `IsValid()` method just for `messageStatus` etc.) ? We already have 150 lines of code just for two enumeration, i think expanding it might be a little bit confusing to future developers of this project. What do you think ?

**Review by aeneasr — DISMISSED** [review]

**Review by aeneasr — DISMISSED** [review]
Great job!
Awesome, thank you! 🎉 Your contribution makes Ory better :)

**Pull request merged**
